src/category_theory/subobject.lean

```diff
@@ -0,0 +1,296 @@
+universes v u
+
+noncomputable theory
+
+namespace category_theory
+
+variables {C : Type u} [category.{v} C]
+
+/-- The category of monomorphisms into `X` is "thin", i.e. a preorder. -/
+instance monomorphisms_subsingleton {X : C} (A B : { f : over X // mono f.hom }) :
+  subsingleton (A ⟶ B) :=
+begin
+  fsplit,
+  rintros ⟨f, _, wf⟩ ⟨g, _, wg⟩,
+  dsimp at *, simp at wf wg, ext, dsimp,
+  exact (@cancel_mono _ _ _ _ _ _ B.property _ _).mp (wf.trans (wg.symm)),
+end
+
+/--
+We define the subobjects of `X` simply to be the isomorphism classes of monomorphisms into `X`.
+See https://ncatlab.org/nlab/show/subobject
+
+One could instead just take the monomorphisms directly: not much changes!
+See https://mathoverflow.net/questions/184196/concise-definition-of-subobjects
+However if we follow this route we only get a preorder, not a partial order,
+which is less convenient for describing lattice properties.
+-/
+def subobject (X : C) := isomorphism_classes.obj (Cat.of { f : over X // mono f.hom })
+
+namespace subobject
+
+/-- Construct a subobject from an explicit monomorphism. -/
+def mk {X Y : C} (f : X ⟶ Y) [w : mono f] : subobject Y :=
+quot.mk _ ⟨over.mk f, w⟩
+
+instance (X : C) : inhabited (subobject X) := ⟨mk (𝟙 _)⟩
+
+/-- The underlying object of a subobject. -/
+def X {X : C} (A : subobject X) : C :=
+(isomorphism_classes.representative A).val.left
+
+/-- The inclusion of a subobject into the ambient object. -/
+def ι {X : C} (A : subobject X) : A.X ⟶ X :=
+(isomorphism_classes.representative A).val.hom
+
+instance {X : C} (A : subobject X) : mono (ι A) :=
+(isomorphism_classes.representative A).property
+
+/-- The underlying object of the subobject constructed from an explicit monomorphism is isomorphic
+to the original source object. -/
+def mk_X_iso {X Y : C} (f : X ⟶ Y) [w : mono f] : (mk f).X ≅ X :=
+(over.forget Y).map_iso
+  ((full_subcategory_inclusion _).map_iso
+  (@isomorphism_classes.mk_representative_iso (Cat.of { f : over Y // mono f.hom }) ⟨over.mk f, w⟩))
+
+@[simp]
+lemma mk_X_iso_hom_comm {X Y : C} (f : X ⟶ Y) [w : mono f] :
+  (mk_X_iso f).hom ≫ f = (mk f).ι :=
+begin
+  have h := ((full_subcategory_inclusion _).map_iso
+    (@isomorphism_classes.mk_representative_iso (Cat.of { f : over Y // mono f.hom })
+      ⟨over.mk f, w⟩)).hom.w,
+  dsimp at h,
+  simpa only [category.comp_id] using h,
+end
+
+@[simp]
+lemma mk_X_iso_id_hom {X : C} : (mk_X_iso (𝟙 X)).hom = (mk (𝟙 X)).ι :=
+by rw [←mk_X_iso_hom_comm, category.comp_id]
+
+@[simp]
+lemma mk_X_iso_inv_comm {X Y : C} (f : X ⟶ Y) [w : mono f] :
+  (mk_X_iso f).inv ≫ (mk f).ι = f :=
+by simp [iso.inv_comp_eq]
+
+/--
+The preorder on subobjects of `X` is `(A,f) ≤ (B,g)`
+if there exists a morphism `h : A ⟶ B` so `h ≫ g = f`.
+(Such a morphism is unique if it exists; in a moment we upgrade this to a `partial_order`.)
+-/
+instance (X : C) : preorder (subobject X) :=
+{ le := λ A B,
+  nonempty (isomorphism_classes.representative A ⟶ isomorphism_classes.representative B),
```

Comment:
I'm pretty this preorder would be easier to work with if the ordering is defined using `quotient.lift_on2` and showing it's independent of the representative, rather than picking a representative; so that if we have two monomorphisms which correspond to the same subobject then we don't need to insert isomorphisms each time.

Reply:
I'll have a think about it (sometime). If there's existing code from your topos project that can/should replace this, I'm also happy to ditch stuff!

Reply:
@b-mehta, my need for this PR has lessened --- could you maybe tell me what will and won't make it more likely that your subobject lattice work makes it into mathlib? I'll then decide what to do here based on that.

(I wanted to talk about the intersection of some kernels, to do the normalized Moore complex of a simplicial module, but I think I'm just going to work elementwise in any case.)